src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,4058 @@
+
+        if input_values is None:
+            raise ValueError(
+                "Both `input_values` and `inputs_embeds` are `None` in `SeamlessM4TSpeechEncoder.forward`. Make sure one of them is not `None`."
+            )
+
+        hidden_states, _ = self.feature_projection(input_values)
+
+        encoder_outputs = self.encoder(
+            hidden_states,
+            attention_mask=attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        hidden_states = encoder_outputs[0]
+
+        # corresponds to UnitYEncoderAdaptor._expand_contract
+        expanded_hidden_states = self.proj1(hidden_states)
+        expanded_hidden_states = self.activation(expanded_hidden_states)
+        expanded_hidden_states = self.proj2(expanded_hidden_states)
+
+        hidden_states = hidden_states + 0.5 * expanded_hidden_states
+
+        if self.adapter is not None:
+            hidden_states = self.adapter(hidden_states, attention_mask=attention_mask)
+
+        hidden_states[0] = self.inner_layer_norm(hidden_states[0])
+
+        if not return_dict:
+            return (hidden_states,) + encoder_outputs[1:]
+
+        return Wav2Vec2BaseModelOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+        )
+
+
+# inspired from MBart and NllbMoe
+class SeamlessM4TEncoder(SeamlessM4TPreTrainedModel):
+    """
+    Transformer encoder consisting of *config.encoder_layers* self attention layers. Each layer is a
+    [`SeamlessM4TEncoderLayer`].
+
+    Args:
+        config: (`SeamlessM4TConfig`)
+        embed_tokens (`nn.Embedding`, *optional*): output embedding
+        is_t2u_encoder (`bool`, *optional*, defaults to `False`):
+            indicates if it belongs to the text-to-units model, in which case it won't have input embeddings
+    """
+
+    def __init__(
+        self,
+        config: SeamlessM4TConfig,
+        embed_tokens: Optional[nn.Embedding] = None,
+        is_t2u_encoder: bool = False,
+    ):
+        super().__init__(config)
+
+        self.dropout = config.dropout
+        self.layerdrop = config.encoder_layerdrop
+        self.padding_idx = config.unit_pad_token_id if is_t2u_encoder else config.pad_token_id
+        embed_dim = config.hidden_size
+        encoder_layers = config.t2u_encoder_layers if is_t2u_encoder else config.encoder_layers
+        encoder_attention_heads = (
+            config.t2u_encoder_attention_heads if is_t2u_encoder else config.encoder_attention_heads
+        )
+        encoder_ffn_dim = config.t2u_encoder_ffn_dim if is_t2u_encoder else config.encoder_ffn_dim
+        self.is_t2u_encoder = is_t2u_encoder
+        self.max_source_positions = config.max_position_embeddings
+
+        if not self.is_t2u_encoder:
```

Comment:
I strongly suspect that it is possible to add embeddings to the `t2u_encoder` which would be the same as the `t2u_decoder` and would allow to do monolingual text-to-speech with the `t2u_model` alone !
In such a case, it would reunite the `t2_encoder` and the `text_encoder`!

I will investigate that once the PR is ready haha